langserver/diagnostics/diagnostics.go

```diff
@@ -75,8 +76,18 @@ func hclDiags(docs <-chan documentContext) {
 		if err := jrpc2.PushNotify(doc.ctx, "textDocument/publishDiagnostics", lsp.PublishDiagnosticsParams{
 			URI:         doc.uri,
 			Diagnostics: hclParse(doc),
-		}); err != nil {
+		}); fatalError(err) {
 			panic(err)
 		}
 	}
 }
+
+func fatalError(err error) bool {
+	if err == nil {
+		return false
+	}
+	if strings.Contains(err.Error(), "client connection is closed") {
```

Comment:
The diagnostics service runs in its own goroutine, so naively I panic if I encounter a fatal error pushing the diagnostics. I have found in testing, there is sometimes a race between shutdown and the diagnostics being sent, but I felt that that error should be nonfatal. I tried `err.(neterr)` and `errors.As` to no avail, this was my naive way of treating the closed connection as nonfatal.

Reply:
I don't think that the LS should panic for any regular known error. Admittedly I didn't catch this in the initial review, but I think that if we're catching errors from notification pushes, then we can just log it as it's not _critical_ - i.e. we shouldn't need to shut down the whole server because we failed to send a notification IMO?

Reply:
Sounds good, I just need to get a logger passed in